### ISSUE TITLE
Cannot cross-build Haskell-OpenGL for Android

### ISSUE THREAD

**Issue by nomeata** [NixOS/nixpkgs#56121]
## Issue description

I’d like to use nix to develop a SDL-based Haskell app for Android, but I can’t cross-compile the Haskell OpenGL library.

### Steps to reproduce

```
~/build/nixpkgs $ nix-build -A pkgsCross.aarch64-android-prebuilt.haskellPackages.OpenGL
error: unsupported platform for Mesa
(use '--show-trace' to show detailed location information)
```

(At least this was the error message when doing this via `reflex-platform`, I cannot actually reproduce this at the moment with `nixpkgs` HEAD,  as it would require downloading 1.7GB, which the current hotel room connection does not really allow.)

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.0-2-amd64, Debian GNU/Linux, noversion`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.0.4`
find: '/nix/var/nix/profiles/per-user': No such file or directory
```

This was originally reported as https://github.com/reflex-frp/reflex-platform/issues/454.

**Comment by nomeata** [comment]
Incidentially, it is possible to cross-compile Haskell SDL and OpenGL for Android, as shown by this project: https://gitlab.com/alleycatcc/haskell-android-sdl
I’d just prefer to use nix than docker :-)

**Comment by nomeata** [comment]
Actually, I can’t run the above for what is likely an unrelated problem with nixpkgs HEAD:
```
[100%] Running all regression tests
/nix/store/954lfsfiv9vhrfgq9r7i52aijnzrhidv-python-2.7.15/bin/python2.7 /tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/./bin/llvm-lit -sv /tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/utils/lit /tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/test
llvm-lit: /tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/utils/lit/tests/lit.cfg:61: warning: Could not import psutil. Some tests will be skipped and the --timeout command line argument will not work.
-- Testing: 23291 tests, 8 threads --
Testing: 0 .. 10..
FAIL: LLVM :: CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll (5120 of 23291)
******************** TEST 'LLVM :: CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll' FAILED ********************
Script:
--
/tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/bin/llc -march=amdgcn -mcpu=verde -verify-machineinstrs < /tmp/nix-build-llvm-6.0.1.drv-0/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll | /tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/bin/FileCheck -check-prefix=GCN -check-prefix=SI /tmp/nix-build-llvm-6.0.1.drv-0/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll
/tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/bin/llc -march=amdgcn -mcpu=tonga -verify-machineinstrs < /tmp/nix-build-llvm-6.0.1.drv-0/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll | /tmp/nix-build-llvm-6.0.1.drv-0/llvm/build/bin/FileCheck -check-prefix=GCN -check-prefix=VI /tmp/nix-build-llvm-6.0.1.drv-0/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll
--
Exit Code: 1

Command Output (stderr):
--
<stdin>:107:40: error: GCN-NOT: string occurred!
 .amd_amdgpu_isa "amdgcn-unknown-linux-android-gfx601"
                                       ^
/tmp/nix-build-llvm-6.0.1.drv-0/llvm/test/CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll:28:12: note: GCN-NOT: pattern specified here
; GCN-NOT: and
           ^

--

********************
Testing: 0 .. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..

1 warning(s) in tests.
Testing Time: 252.31s
********************
Failing Tests (1):
    LLVM :: CodeGen/AMDGPU/llvm.amdgcn.mbcnt.ll

  Expected Passes    : 22189
  Expected Failures  : 138
  Unsupported Tests  : 963
  Unexpected Failures: 1
make[3]: *** [CMakeFiles/check-all.dir/build.make:58: CMakeFiles/check-all] Error 1
make[3]: Leaving directory '/tmp/nix-build-llvm-6.0.1.drv-0/llvm/build'
make[2]: *** [CMakeFiles/Makefile2:361: CMakeFiles/check-all.dir/all] Error 2
make[2]: Leaving directory '/tmp/nix-build-llvm-6.0.1.drv-0/llvm/build'
make[1]: *** [CMakeFiles/Makefile2:368: CMakeFiles/check-all.dir/rule] Error 2
make[1]: Leaving directory '/tmp/nix-build-llvm-6.0.1.drv-0/llvm/build'
make: *** [Makefile:251: check-all] Error 2
builder for '/nix/store/x461sarmy3rw3k8v3dr7p7rx82ycgai8-llvm-6.0.1.drv' failed with exit code 2
```

**Comment by nomeata** [comment]
LLVM bug reported separately as #56156.

**Comment by matthewbauer** [comment]
You probably don’t want to build mesa on android. They aren’t going to work as well as the native driver. I think the ndk has its own opengl you can link agaibst?

**Comment by nomeata** [comment]
Sorry, but this is lower than my usual comfort zone, as I usually deal with Haskell libraries. Do you mean that the Haskell libraries should depend on something else?

**Comment by matthewbauer** [comment]
I’m not 100% sure but i suspect that haskell opengl does not work with the embedded versions of OpenGL provided by android. You might be able to use this one though:

http://hackage.haskell.org/package/opengles

**Comment by nomeata** [comment]
In principle it can be made to work, as this docker-based project shows: https://gitlab.com/alleycatcc/haskell-android-sdl (the [demo](https://gitlab.com/alleycatcc/sdl-gles-cairo-demo/blob/master/sdl-gles-cairo-demo.cabal) depends on the Haskell `OpenGL` version)…

ah, and there I find an interesting bit of information: The Haskell OpenGLRaw package has a flag `-fosandroid` that ought to be enabled on android.

There is also `-fusegles2` – “Use OpenGL ES 2.x instead of OpenGL ES 3.x on Android”

The file https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/hackage-packages.nix does not make any provision to set that flag. So maybe that would help.

@peti, you seem to be an expert on this: Is there a way to set these flags correctly when building for Android? Would this prevent the android build from pulling in Mesa?

**Comment by matthewbauer** [comment]
Oh ok. Yeah it looks like it should work. The headers and libraries should be provided by the ndk. Let me try to build it.

**Comment by peti** [comment]
> Is there a way to set these flags correctly when building for Android?

There are two functions, `pkgs.haskell.lib.enableCabalFlag` and `pkgs.haskell.lib.disableCabalFlag`, that you can use to create a variant of the default `OpenGLRaw` package with the appropriate flags enabled or disabled. You may also have to override the `Mesa` library that's being passed to the build function to some appropriate version. Check out the file `pkgs/development/haskell-modules/lib.nix`, which has lots of useful helpers to do this. https://nixos.org/nixpkgs/manual/#how-to-build-projects-that-depend-on-each-other shows how to configure those overrides in your `~/.config/nixpkgs/config.nix`.

Generally speaking, I have not been able to re-produce the error you've mentioned in your initial post. 

    $ nix-build "<nixpkgs>" -A pkgsCross.aarch64-android-prebuilt.haskellPackages.OpenGL

gives no evaluation errors when I run it on a current version of Nixpkgs `master`.

**Comment by nomeata** [comment]
> … gives no evaluation errors when I run it on a current version of Nixpkgs `master`.

indeed, it no longer does that here either. Must have been related to `reflex-platform` somehow, or the version of nixpkgs pinned there. So ignore that part, sorry.

So may this all works once #56156 is fixed?

**Comment by nomeata** [comment]
#56156 has a fix at #56272 but now this is blocked by #56274.
